classes/helper/mahara_portfolio.php

```diff
@@ -0,0 +1,273 @@
+        // but for now we have to save to a temp file and then extract it.
+        $url = $this->fileurl . '&wstoken=' . $this->wstoken;
+
+        $fs = get_file_storage();
+        $this->zipfile = $fs->create_file_from_url($this->fileinfo, $url);
+    }
+
+    /**
+     * Extract files from zip and add to the Ouriginal queue.
+     *
+     */
+    private function queue_files() {
+        global $CFG;
+        mtrace('urkund: mahara_portfolio: Extracting files from zip file.');
+        require_once($CFG->dirroot . '/plagiarism/urkund/lib.php');
+
+        $success = true;
+        $zipfile = $this->zipfile;
+        $fileinfo = $this->fileinfo;
+        $dirpath = $fileinfo['filepath'] . 's' . $this->submission->id . '/';
+        if ($this->relateduserid) {
+            $theuserid = $this->relateduserid;
+        } else {
+            $theuserid = $this->userid;
+        }
+
+        // Extract the files to storage and save them as Moodle files,
+        // to make them available to the Ouriginal queue in a scheduled task.
+        $packer = new \zip_packer();
+        $zipcontents = $zipfile->extract_to_storage($packer, $this->context->id, $fileinfo['component'],
+                $fileinfo['filearea'], $this->submission->submission, $dirpath, $theuserid);
+        if (!$zipcontents) {
+            mtrace('urkund: mahara_portfolio: Extracting zip failed - aborting');
+            return false;
+        }
+        $fs = get_file_storage();
+        foreach (array_keys($zipcontents) as $entry) {
+            if ($pos = strpos($entry, '/')) {
+                // Split path components from the entry into the path param.
+                $thisdir = $dirpath . substr($entry, 0, $pos + 1);
+                $thisfile = substr($entry, $pos +1 );
+            } else {
+                $thisdir = $dirpath;
+                $thisfile = $entry;
+            }
+
+            // Load as a Moodle file object, suitable for adding to the queue.
+            $file = $fs->get_file($this->context->id, $fileinfo['component'], $fileinfo['filearea'],
+                    $this->submission->submission, $thisdir, $thisfile);
+            if ($file->get_filename() != '.') {
+                mtrace('urkund: mahara_portfolio: Queue file: ' . $entry);
+                $result = urkund_queue_file($this->instanceid, $this->userid, $file, $this->relateduserid);
+                if (!$result) {
+                    $success = false;
+                    mtrace('urkund: mahara_portfolio: Queue file failed - aborting: ' . $entry);
+                }
+            }
+        }
+
+        if ($success) {
+            global $DB;
+            // Remove this submission's maharawsdl entry from queue.
+            $DB->delete_records('plagiarism_urkund_files', array(
+                'cm' => $this->instanceid,
+                'userid' => $this->userid,
+                'statuscode' => 'maharawsdl'
+            ));
+            // Remove saved zip file from storage.
```

Comment:
This feels like it could leave a bunch of un-used zip files I'm not sure we need to wrap this in a call to if ($success) unless debugging - for example - if urkund_queue_file returns '' because the file isn't one we care about etc.